---
title: "Two Donor scRNAseq"
output:
  html_document:
    df_print: paged
knit: (function(inputFile, encoding) {
  rmarkdown::render(inputFile, encoding = encoding, output_dir = here::here('..','html'))})
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(dplyr)
library(tidyr)
library(tidyr)
library(Seurat)
library(patchwork)
library(cluster)
library(parallelDist)
library(ggplot2)
library(data.table)
library(cowplot)
library(future)
library(SeuratDisk)


reload_data=F
source(here::here('r','fig_colors.R'))
source(here::here('r','scrna','HTOdemux.R'))
source(here::here('r','scrna','scrna_analysis_fxns.R'))

options(future.globals.maxSize = 14000 * 1024^2)
plan("sequential")
plan()
```

## Load all four wells individually

```{r}

#rename some adt features
hto_rows <- paste('Hashtag',1:12,sep='_')

adt_feature_names <- fread(here::here(
    '..','..','scrnaseq_tcsl154',
    'meta','adt_names.tsv'))

make_sample_seurat_obj <- function(expt_num, well) {

  scrna.data <- Read10X(data.dir = here::here(
    '..', '..', paste0('scrnaseq_tcsl',expt_num),
    'cellranger', paste('TCSL', expt_num, well, sep='_'), 'outs','raw_feature_bc_matrix'))

  rownames(scrna.data$`Antibody Capture`) <- c(paste0(adt_feature_names$vis_name,'.adt'), hto_rows)
  
  #rna features
  scrna <- CreateSeuratObject(
      counts = scrna.data$`Gene Expression`, 
      project = "tcsl154", 
      min.cells = 3,
      min.features = 200)
  
  adt_rows <- setdiff(rownames(scrna.data$`Antibody Capture`), hto_rows)
  
  #adt features
  scrna[['ADT']] <- CreateAssayObject(
      counts = scrna.data$`Antibody Capture`[adt_rows, colnames(scrna)])
  
  #hto features
  scrna[['HTO']] <- CreateAssayObject(
    counts = scrna.data$`Antibody Capture`[hto_rows, colnames(scrna)])
  
  scrna@meta.data$sample <- paste(expt_num, well, sep='_')
  scrna@meta.data$donor <- expt_num
  scrna@meta.data$well <- well
  
  return(scrna)
}

scrna.list <- list(
  make_sample_seurat_obj(154,1),
  make_sample_seurat_obj(154,2),
  make_sample_seurat_obj(170,1),
  make_sample_seurat_obj(170,2)
)
```



## Remove MT-high cells and individually perform SCTransform on all 3 assay types.

```{r}
mt_rna_pct_cutoff <- 25
read_cutoff <- 1e3

run_sct <- function(seurat_obj, mt_rna_pct_cutoff, read_cutoff) {
  
  # remove high MT-RNA cells (dying/dead?)
  DefaultAssay(seurat_obj) <- 'RNA'
  seurat_obj <- PercentageFeatureSet(seurat_obj, pattern = "^MT-", col.name = "percent.mt")
  seurat_obj <- subset(seurat_obj, subset=(percent.mt < mt_rna_pct_cutoff & nCount_RNA > read_cutoff))
  
  # get cell cycle scores
  DefaultAssay(seurat_obj) <- 'RNA'
  s.genes <- cc.genes$s.genes
  g2m.genes <- cc.genes$g2m.genes
  #https://satijalab.org/seurat/v3.1/cell_cycle_vignette.html
  seurat_obj <- CellCycleScoring(seurat_obj, s.features = s.genes, g2m.features = g2m.genes, set.ident = TRUE)
  #seurat_obj <- RunPCA(seurat_obj, features = c(s.genes, g2m.genes), reduction.name = 'cc_pca')
  message('  Cell Cycle Scoring Complete.')
  
  #SCTransform for RNA
  DefaultAssay(seurat_obj) <- 'RNA'
  seurat_obj <- SCTransform(seurat_obj, 
    vars.to.regress = c("S.Score", "G2M.Score"), verbose = FALSE)
  seurat_obj <- RunPCA(seurat_obj, reduction.name = 'rna_pca')
  message('  SCTransform RNA complete.')
  
  #SCTransform for ADT
  DefaultAssay(seurat_obj) <- 'ADT'
  seurat_obj <- SCTransform(
    seurat_obj, assay='ADT', new.assay.name = 'SCT_ADT', 
    vars.to.regress = c("S.Score", "G2M.Score"), verbose = FALSE)
  seurat_obj <- RunPCA(
    seurat_obj, features = rownames(seurat_obj[["SCT_ADT"]]), 
    nfeatures.print = 10, reduction.name='adt_pca')
  message('  SCTransform ADT complete.')
  
  #SCTransform for HTO
  DefaultAssay(seurat_obj) <- 'HTO'
  seurat_obj <- SCTransform(seurat_obj, assay='HTO', new.assay.name = 'SCT_HTO')
  message('  SCTransform HTO complete.')
  
  return(seurat_obj)
}

for (i in 1:4) {
  message(paste('Starting Sample',i))
  scrna.list[[i]] <- run_sct(scrna.list[[i]], mt_rna_pct_cutoff, read_cutoff)
   message('  Sample completed.')
}

initial.rds.file <- here::here(
     '..','..','scrnaseq_tcsl154',
     'rds',
     paste0('combined.initial.data.rds'))

#if (!file.exists(initial.rds.file)) {
saveRDS(scrna.list,initial.rds.file)

```

```{r}


combined_metadata <- rbindlist(lapply(1:4, function(i) data.table(scrna.list[[i]]@meta.data)))
mt_summary <- combined_metadata[,
  mt_thresh := percent.mt < mt_rna_pct_cutoff & nCount_RNA > read_cutoff][,
  list(
    neg= .N-sum(mt_thresh),
    pos=sum(mt_thresh),
    pct_pos=round(sum(mt_thresh)/.N, 2)), by=c('donor')]

ggplot(combined_metadata) + 
  aes(x=nCount_RNA, y=percent.mt) + 
  scale_x_log10() + 
  stat_density2d(geom="tile", aes(fill=..density.., alpha=1), contour=FALSE) + 
  geom_point(size=0.5, color=blues9[9], alpha=0.25) +
  stat_density2d(geom="tile", aes(fill=..density..,     alpha=ifelse(..density..<0.01,0,1)), contour=FALSE) + 
  scale_fill_gradientn(colours = colorRampPalette(c("white", blues9))(256)) +
  facet_grid(~donor) +
  geom_hline(aes(yintercept=mt_rna_pct_cutoff)) +
  geom_vline(aes(xintercept=read_cutoff)) +
  geom_text(data=mt_summary, aes(x=10^4,y=100, label=paste(pct_pos,":",pos,'kept,',neg,'removed'))) +
  theme_bw()
```

Percent MT-RNA is higher for 170, and more cells are lost. May want to consider a higher % cutoff for that sample, but
this is already quite permissive according to some literature (10% is considered optimal by Osorio et al. 2020).

### HTO calling for each individual well

Getting unclear error messages with HTO SCT integration, so calling cells separately for each well, then combining
via metadata.

```{r}

HTO_sample_metadata <- fread(here::here(
    '..','..','scrnaseq_tcsl154','meta','sample_metadata.csv'))

for (i in 1:4) {
  DefaultAssay(scrna.list[[i]]) <- 'SCT_HTO'
  
  if (!('hto.tsne' %in% names(scrna.list[[i]]@reductions))) {
    scrna.list[[i]] <- RunTSNE(
      scrna.list[[i]],
      perplexity=100, dims=NULL,
      features=paste0('Hashtag-',1:12),
      reduction.name='hto.tsne', tsne.method='FIt-SNE', nthreads=12)
  }
  
  donor_i <- ifelse(i < 3, 'TCSL154','TCSL170')
  sample_metadata <- HTO_sample_metadata[donor == donor_i]
  sample_metadata[,  sample := paste(car, ifelse(k_type=='baseline', 'unstim','stim'), sep='.')]
  sample_metadata[k_type=='bead_stim',  sample := paste(car, k_type, sep='.')]
  
  scrna.list[[i]]@meta.data$well <- scrna.list[[i]]@meta.data$sample
  scrna.list[[i]]@meta.data$sample <- NULL
  
  scrna.list[[i]] <- assign_HTO_sample(
    scrna.list[[i]], metadata=sample_metadata,
    assay='SCT_HTO', slot='data', save_plots=T)

}

plots <- list()

plots$well_tsne_maps <- plot_grid(
  plotlist=unlist(
    lapply(1:4, function (i) list(
      DimPlot(scrna.list[[i]], group.by='sample', reduction='hto.tsne'),
      DimPlot(scrna.list[[i]], group.by='HTO_status', reduction='hto.tsne'))), 
    recursive=F), ncol=2)

plots$well_tsne_maps_pruned <- plot_grid(
  plotlist=unlist(
    lapply(1:4, function (i) {
      scrna_subset <- subset(scrna.list[[i]], subset = (HTO_status %in% c('HighQual','MidQual')))
      list(
      DimPlot(scrna_subset, group.by='sample', reduction='hto.tsne'),
      DimPlot(scrna_subset, group.by='HTO_status', reduction='hto.tsne'))
    }), 
    recursive=F), 
  ncol=2)

```

### Integrate SCT RNA

```{r}

integrate_feature <- function(scrna.list, assay_name, initial_assay, assay_only=F, remove_other_assays=F, use_ref=F) {
  
  integrated_assay_name <- paste0(assay_name, "_INT")
  
  # only one assay
  if (remove_other_assays) {
    scrna.assay.list <- lapply(
      1:length(scrna.list),
      function(i) {
        obj <- CreateSeuratObject(
          scrna.list[[i]][[assay_name]],
          assay=assay_name,
          meta.data=scrna.list[[i]]@meta.data)
        obj[[initial_assay]] <- scrna.list[[i]][[initial_assay]]
        return(obj)
      }
    )
  } else {
    scrna.assay.list <- scrna.list
  }
  
  for (i in 1:length(scrna.assay.list)) {
    DefaultAssay(scrna.assay.list[[i]]) <- assay_name
  }
  
  integrate.features <- SelectIntegrationFeatures(
    object.list = scrna.assay.list, 
    nfeatures = 3000,
    assay=rep(assay_name, length(scrna.assay.list)))
  gc()
  
  scrna.assay.list <- PrepSCTIntegration(
    assay=assay_name, object.list = scrna.assay.list, 
    anchor.features = integrate.features, 
    verbose = TRUE)
  gc()
  
  if (use_ref) {
    scrna.anchors <- FindIntegrationAnchors(
      object.list = scrna.assay.list,
      normalization.method = "SCT",
      anchor.features = integrate.features,
      reference = use_ref)
  } else {
    scrna.anchors <- FindIntegrationAnchors(
      object.list = scrna.assay.list,
      normalization.method = "SCT", 
      anchor.features = integrate.features,
      verbose = TRUE)
  }
  gc()
  
  scrna.integrated <- IntegrateData(
    anchorset = scrna.anchors,
    normalization.method = "SCT", 
    verbose = TRUE,
    new.assay.name = integrated_assay_name)
  gc()
    
  if (assay_only) {
    return(scrna.integrated[[integrated_assay_name]])
  } else {
    return(scrna.integrated)
  }
}

scrna.integrated <- integrate_feature(
  scrna.list, 
  assay_name='SCT',
  initial_assay='RNA',
  assay_only=F,
  remove_other_assays=F)
scrna.integrated[['SCT_ADT_INT']] <- integrate_feature(
  scrna.list,
  assay_name='SCT_ADT',
  initial_assay='ADT',
  assay_only=T,
  remove_other_assays=T)

integrated.rds.file <- here::here(
     '..','..','scrnaseq_tcsl154',
     'rds',
     paste0('combined.integrated.umap.data.rds'))

saveRDS(scrna.integrated,integrated.rds.file)
```

### Prune HTO and CD4/CD8 Doublets, assign CD4/CD8

```{r}

scrna.integrated <- subset(scrna.integrated, subset = (HTO_status %in% c('HighQual','MidQual')))
hto_subset <- subset(scrna.integrated, cells=sample(Cells(scrna.integrated), 8000))

t_out <- reassign_t_type(scrna.integrated, n_clust=3, plots)
plots <- t_out$plot_obj
scrna.integrated <- t_out$seurat_obj
rm(t_out)

plots$t_type_bars <- ggplot(
  data.table(scrna.integrated@meta.data)) +
  geom_bar(aes(x=car, fill=t_type)) +
  facet_grid(k_type~donor, space='free', scales='free') +
  coord_flip()
```
## Run WNN Clustering on all cells

```{r}

s.genes <- cc.genes$s.genes
g2m.genes <- cc.genes$g2m.genes

scrna.integrated <- RunPCA(scrna.integrated, features = c(s.genes, g2m.genes), reduction.name = 'cc_pca')
scrna.integrated <- RunPCA(scrna.integrated, assay='SCT_INT', reduction.name = 'rna_pca')

# identify unique TCSL170 ADTs
adt_count_ratios <- (
  Matrix::rowSums(subset(scrna.integrated, subset=(donor == 'TCSL154'))[['SCT_ADT_INT']]@counts) / 
    Matrix::rowSums(subset(scrna.integrated, subset=(donor == 'TCSL170'))[['SCT_ADT_INT']]@counts))
tcsl170_adts <- names(which(-log10(adt_count_ratios) > 2))
shared_adts <- names(which(-log10(adt_count_ratios) < 2))

scrna.integrated <- RunPCA(scrna.integrated, assay='SCT_ADT_INT', features= shared_adts, 
    nfeatures.print = 10, reduction.name='adt_pca')
scrna_output <- sct_umap(scrna.integrated, title='All cells')

scrna.integrated <- scrna_output$obj
plots <- c(plots, scrna_output$plots)
rm(scrna_output)

plots$combined_umap_2 <- (
  DimPlot(scrna.integrated, reduction = 'wnn.umap', label = TRUE, repel = TRUE, label.size = 2.5) + 
    NoLegend() +
  DimPlot(scrna.integrated, reduction = 'wnn.umap', group.by = 't_type', label = TRUE, repel = TRUE, label.size = 2.5) + 
    NoLegend() +
  DimPlot(scrna.integrated, reduction = 'wnn.umap', group.by = 'k_type', label = TRUE, repel = TRUE, label.size = 2.5) +
    NoLegend() +
  DimPlot(scrna.integrated, reduction = 'wnn.umap', group.by = 'Phase', label = TRUE, repel = TRUE, label.size = 2.5) + 
    NoLegend() +
  DimPlot(scrna.integrated, reduction = 'wnn.umap', group.by = 'donor', label = TRUE, repel = TRUE, label.size = 2.5) + 
    NoLegend())
```


## Map TCSL170 subtypes onto new UMAP

```{r}

scrna.wnn.cd4_all_metadata <- readRDS("~/scrnaseq_tcsl154/rds/scrna.wnn.cd4_all_metadata.rds")
scrna.wnn.cd8_all_metadata <- readRDS("~/scrnaseq_tcsl154/rds/scrna.wnn.cd8_all_metadata.rds")

original_metadata <- rbind(
    data.table(scrna.wnn.cd4_all_metadata, keep.rownames=T)[, list(rn_154=rn, sample, t_type, Subtype)],
    data.table(scrna.wnn.cd8_all_metadata, keep.rownames=T)[, list(rn_154=rn, sample, t_type, Subtype)])[,
      rn := gsub('(.*)-\\d','\\1',rn_154)]

integrated_metadata <- data.table(
  scrna.integrated@meta.data, keep.rownames=T)[,
    list(rn_orig=rn, sample, donor, orig_t_type=t_type)][
      donor == 'TCSL154', rn := gsub('(.*)-\\d_\\d','\\1',rn_orig)]

sample_subtype_map <- merge(integrated_metadata, original_metadata, by='rn', all=T)[!is.na(Subtype) & !is.na(rn_orig)]

scrna.integrated@meta.data[,'subtype_154'] <- factor(NA, levels=levels(sample_subtype_map$Subtype))
scrna.integrated@meta.data[sample_subtype_map$rn_orig,]$subtype_154 <- sample_subtype_map$Subtype

DimPlot(scrna.integrated, reduction = 'wnn.umap', group.by = 'subtype_154', label = TRUE, repel = TRUE, label.size = 2.5)
```

## Donor, T, and Stim amounts per cluster
```{r}
ggplot(data.table(scrna.integrated@meta.data)[t_type != 'DP'][, 
    .N, by=c('seurat_clusters','donor','k_type','t_type')][N > 100]) + 
  geom_bar(aes(x=seurat_clusters, y=N, fill=interaction(donor,k_type)), stat='identity') +
  facet_wrap(~t_type, scales='free')

## END OF DATA LOADING

```
------------------------------------------------------------------------------------------------------------------------

Finally, ran UMAP creation for all subsets made HD5 seurat objects for each with `07_run_subset_job_indiv.r`.

------------------------------------------------------------------------------------------------------------------------

## Map new clusters onto old subtypes


Easiest way is to match based on most frequently assigned subtype. Do CD8 only first.

```{r}
scrna_cd8 <- LoadH5Seurat(here::here(
  '..','..','scrnaseq_tcsl154',
  'rds',
  paste0('scrna.wnn.int.CD8.act_clust.clean.h5Seurat')))

int_cd8_subset <- subset(combined.integrated.data,
    cells=intersect(
      names(scrna_cd8$cell), 
      names(combined.integrated.data$cell)))

scrna_cd8[['RNA']] <- int_cd8_subset@assays$RNA
scrna_cd8[['ADT']] <- int_cd8_subset@assays$ADT

## phase plots

phase_plots <- (
  FeaturePlot(scrna_cd8, features=c('G2M.Score'), min.cutoff=-2, max.cutoff=1)+
  FeaturePlot(scrna_cd8, features=c('S.Score'), min.cutoff=-1.5, max.cutoff=1)+
  DimPlot(scrna_cd8, group.by ='Phase'))

cd8_metadata <- data.table(
  scrna_cd8@meta.data, keep.rownames=T)[,
    list(rn_orig=rn, sample, donor, orig_t_type=t_type)][
      donor == 'TCSL154', rn := gsub('(.*)-\\d_\\d','\\1',rn_orig)]

scrna.wnn.cd4_all_metadata <- readRDS("~/scrnaseq_tcsl154/rds/scrna.wnn.cd4_all_metadata.rds")
scrna.wnn.cd8_all_metadata <- readRDS("~/scrnaseq_tcsl154/rds/scrna.wnn.cd8_all_metadata.rds")

original_metadata <- rbind(
    data.table(scrna.wnn.cd4_all_metadata, keep.rownames=T)[, list(rn_154=rn, sample, t_type, Subtype)],
    data.table(scrna.wnn.cd8_all_metadata, keep.rownames=T)[, list(rn_154=rn, sample, t_type, Subtype)])[,
      rn := gsub('(.*)-\\d','\\1',rn_154)]

sample_subtype_map <- merge(cd8_metadata, original_metadata, by='rn', all=T)[!is.na(Subtype) & !is.na(rn_orig)]
get_cluster_enrich_plot(scrna_cd8, 'seurat_clusters', hide='', facet_formula=formula(~car))

scrna_cd8@meta.data[,'subtype_154'] <- factor(NA, levels=levels(sample_subtype_map$Subtype))
scrna_cd8@meta.data[sample_subtype_map$rn_orig,]$subtype_154 <- sample_subtype_map$Subtype

DimPlot(scrna_cd8, reduction = 'wnn.umap', group.by = 'subtype_154', label = TRUE, repel = TRUE, label.size = 2.5) +
  DimPlot(scrna_cd8, reduction = 'wnn.umap', group.by = 'seurat_clusters', label = TRUE, repel = TRUE, label.size = 2.5)

# cd8_umap_plots <- umap_plot(scrna_cd8, assay='SCT_INT', cluster_name='seurat_clusters', reduction='wnn.umap')
# cd8_umap_plots_adt <- umap_plot(scrna_cd8, assay='SCT_ADT_INT', cluster_name='seurat_clusters', reduction='wnn.umap')

clust_mapping_154 <- compare_clusters(scrna_cd8, 'seurat_clusters', 'subtype_154')
plot_grid(clust_mapping_154$plot_a_in_b, clust_mapping_154$plot_b_in_a)

ggplot(data.table(scrna_cd8@meta.data)[, 
    list(N=.N, seurat_clusters), by=c('donor','car','k_type')][, 
      pct := .N/N, by=c('donor','car','seurat_clusters','k_type')]) + 
  geom_bar(aes(x=seurat_clusters, fill=donor, y=pct), stat='identity', position='dodge') + 
  facet_grid(car + k_type ~ .)
```

### Sheih Et all IRF Gene Signature

```{r}

# Genes from IRF cluster in Fig 7b of Shieh et al 2020 

scrna_cd8 <- AddModuleScore(scrna_cd8, features=list('1'=c(
  'CCL4','CCL5','CD27','CD74','FGFBP2','GNLY','GZMA','GZMB',  
  'GZMH','GZMK','HLA-DRA','HLA-DRB1','HLA-DRB5','IFNG','LAG3',
  'NKG7','STMN1')),name='Sheih_et_al_IRF')

#oxphos hallmark set
oxphos_genes <- c('ABCB7', 'ACAA1', 'ACAA2', 'ACADM', 'ACADSB', 'ACADVL', 'ACAT1', 'ACO2', 'AFG3L2', 'AIFM1', 'ALAS1', 'ALDH6A1', 'ATP1B1', 'ATP5F1A', 'ATP5F1B', 'ATP5F1C', 'ATP5F1D', 'ATP5F1E', 'ATP5MC1', 'ATP5MC2', 'ATP5MC3', 'ATP5ME', 'ATP5MF', 'ATP5MG', 'ATP5PB', 'ATP5PD', 'ATP5PF', 'ATP5PO', 'ATP6AP1', 'ATP6V0B', 'ATP6V0C', 'ATP6V0E1', 'ATP6V1C1', 'ATP6V1D', 'ATP6V1E1', 'ATP6V1F', 'ATP6V1G1', 'ATP6V1H', 'BAX', 'BCKDHA', 'BDH2', 'CASP7', 'COX10', 'COX11', 'COX15', 'COX17', 'COX4I1', 'COX5A', 'COX5B', 'COX6A1', 'COX6B1', 'COX6C', 'COX7A2', 'COX7A2L', 'COX7B', 'COX7C', 'COX8A', 'CPT1A', 'CS', 'CYB5A', 'CYB5R3', 'CYC1', 'CYCS', 'DECR1', 'DLAT', 'DLD', 'DLST', 'ECH1', 'ECHS1', 'ECI1', 'ETFA', 'ETFB', 'ETFDH', 'FDX1', 'FH', 'FXN', 'GLUD1', 'GOT2', 'GPI', 'GPX4', 'GRPEL1', 'HADHA', 'HADHB', 'HCCS', 'HSD17B10', 'HSPA9', 'HTRA2', 'IDH1', 'IDH2', 'IDH3A', 'IDH3B', 'IDH3G', 'IMMT', 'ISCA1', 'ISCU', 'LDHA', 'LDHB', 'LRPPRC', 'MAOB', 'MDH1', 'MDH2', 'MFN2', 'MGST3', 'MPC1', 'MRPL11', 'MRPL15', 'MRPL34', 'MRPL35', 'MRPS11', 'MRPS12', 'MRPS15', 'MRPS22', 'MRPS30', 'MTRF1', 'MTRR', 'MTX2', 'NDUFA1', 'NDUFA2', 'NDUFA3', 'NDUFA4', 'NDUFA5', 'NDUFA6', 'NDUFA7', 'NDUFA8', 'NDUFA9', 'NDUFAB1', 'NDUFB1', 'NDUFB2', 'NDUFB3', 'NDUFB4', 'NDUFB5', 'NDUFB6', 'NDUFB7', 'NDUFB8', 'NDUFC1', 'NDUFC2', 'NDUFS1', 'NDUFS2', 'NDUFS3', 'NDUFS4', 'NDUFS6', 'NDUFS7', 'NDUFS8', 'NDUFV1', 'NDUFV2', 'NNT', 'NQO2', 'OAT', 'OGDH', 'OPA1', 'OXA1L', 'PDHA1', 'PDHB', 'PDHX', 'PDK4', 'PDP1', 'PHB2', 'PHYH', 'PMPCA', 'POLR2F', 'POR', 'PRDX3', 'RETSAT', 'RHOT1', 'RHOT2', 'SDHA', 'SDHB', 'SDHC', 'SDHD', 'SLC25A11', 'SLC25A12', 'SLC25A20', 'SLC25A3', 'SLC25A4', 'SLC25A5', 'SLC25A6', 'SUCLA2', 'SUCLG1', 'SUPV3L1', 'SURF1', 'TCIRG1', 'TIMM10', 'TIMM13', 'TIMM17A', 'TIMM50', 'TIMM8B', 'TIMM9', 'TOMM22', 'TOMM70', 'UQCR10', 'UQCR11', 'UQCRB', 'UQCRC1', 'UQCRC2', 'UQCRFS1', 'UQCRH', 'UQCRQ', 'VDAC1', 'VDAC2', 'VDAC3')

glycolysis_genes <- c('ABCB6', 'ADORA2B', 'AGL', 'AGRN', 'AK3', 'AK4', 'AKR1A1', 'ALDH7A1', 'ALDH9A1', 'ALDOA', 'ALDOB', 'ALG1', 'ANG', 'ANGPTL4', 'ANKZF1', 'ARPP19', 'ARTN', 'AURKA', 'B3GALT6', 'B3GAT1', 'B3GAT3', 'B3GNT3', 'B4GALT1', 'B4GALT2', 'B4GALT4', 'B4GALT7', 'BIK', 'BPNT1', 'CACNA1H', 'CAPN5', 'CASP6', 'CD44', 'CDK1', 'CENPA', 'CHPF', 'CHPF2', 'CHST1', 'CHST12', 'CHST2', 'CHST4', 'CHST6', 'CITED2', 'CLDN3', 'CLDN9', 'CLN6', 'COG2', 'COL5A1', 'COPB2', 'CTH', 'CXCR4', 'CYB5A', 'DCN', 'DDIT4', 'DEPDC1', 'DLD', 'DPYSL4', 'DSC2', 'ECD', 'EFNA3', 'EGFR', 'EGLN3', 'ELF3', 'ENO1', 'ENO2', 'ERO1A', 'EXT1', 'EXT2', 'FAM162A', 'FBP2', 'FKBP4', 'FUT8', 'G6PD', 'GAL3ST1', 'GALE', 'GALK1', 'GALK2', 'GAPDHS', 'GCLC', 'GFPT1', 'GLCE', 'GLRX', 'GMPPA', 'GMPPB', 'GNE', 'GNPDA1', 'GOT1', 'GOT2', 'GPC1', 'GPC3', 'GPC4', 'GPR87', 'GUSB', 'GYS1', 'GYS2', 'HAX1', 'HDLBP', 'HK2', 'HMMR', 'HOMER1', 'HS2ST1', 'HS6ST2', 'HSPA5', 'IDH1', 'IDUA', 'IER3', 'IGFBP3', 'IL13RA1', 'IRS2', 'ISG20', 'KDELR3', 'KIF20A', 'KIF2A', 'LCT', 'LDHA', 'LDHC', 'LHPP', 'LHX9', 'MDH1', 'MDH2', 'ME1', 'ME2', 'MED24', 'MERTK', 'MET', 'MIF', 'MIOX', 'MPI', 'MXI1', 'NANP', 'NASP', 'NDST3', 'NDUFV3', 'NOL3', 'NSDHL', 'NT5E', 'P4HA1', 'P4HA2', 'PAM', 'PAXIP1', 'PC', 'PDK3', 'PFKFB1', 'PFKP', 'PGAM1', 'PGAM2', 'PGK1', 'PGLS', 'PGM2', 'PHKA2', 'PKM', 'PKP2', 'PLOD1', 'PLOD2', 'PMM2', 'POLR3K', 'PPFIA4', 'PPIA', 'PPP2CB', 'PRPS1', 'PSMC4', 'PYGB', 'PYGL', 'QSOX1', 'RARS1', 'RBCK1', 'RPE', 'RRAGD', 'SAP30', 'SDC1', 'SDC2', 'SDC3', 'SDHC', 'SLC16A3', 'SLC25A10', 'SLC25A13', 'SLC35A3', 'SLC37A4', 'SOD1', 'SOX9', 'SPAG4', 'SRD5A3', 'STC1', 'STC2', 'STMN1', 'TALDO1', 'TFF3', 'TGFA', 'TGFBI', 'TKTL1', 'TPBG', 'TPI1', 'TPST1', 'TSTA3', 'TXN', 'UGP2', 'VCAN', 'VEGFA', 'VLDLR', 'XYLT2', 'ZNF292')

scrna_cd8 <- AddModuleScore(scrna_cd8, assay='RNA', features=list('1'=oxphos_genes),name='hallmark_oxphos')

oxphos_gene_clust <- heatmap_clust_v_genes(scrna_cd8, 'seurat_clusters', assay='RNA', features=oxphos_genes, use_pct=T, use_slot='data')$plot

```

### Cluster PCAs

```{r}

  obj.avg <- AverageExpression(subset(
    scrna_cd8, subset=(seurat_clusters %in% c(0:14))),
    assay=c('SCT_INT','SCT_ADT_INT'), add.ident='donor', slot='scale.data')
  obj.avg <- t(rbind(obj.avg$SCT_ADT_INT, obj.avg$SCT_INT))
  obj.avg <- obj.avg[, apply(obj.avg,2,sd) > 0]
  combined_pca <- prcomp(scale(obj.avg))
  
  # calculate pca stats
  pca.dt <- data.table(pc = data.table(colnames(combined_pca$rotation))[, 
      PC := as.integer(gsub("[A-Z]", "", V1))][, PC], 
      sd = combined_pca$sdev, 
      var = combined_pca$sdev^2, 
      var.norm = combined_pca$sdev^2/sum(combined_pca$sdev^2), 
      var.acc = cumsum(combined_pca$sdev^2/sum(combined_pca$sdev^2)))
  
  melt.pca.dt <- melt(
    pca.dt, measure.vars = c("var.norm", "var.acc"), variable.name = "metric")
  
  pca_metrics_plot <- ggplot(melt.pca.dt) +
    geom_line(aes(x = pc, y = value, color = metric)) +
    geom_point(aes(x = pc, y = value, color = metric)) +
    scale_x_continuous(limits = c(1, NA)) +
    labs(title = "Fraction of Variance Captured by Principal Components, Every Sort Group", 
         x = "Principal Component", y = "Fraction of Variance") +
    theme_bw()
  
  projected.metrics <- data.table(
    scale(obj.avg, combined_pca$center, combined_pca$scale) %*% combined_pca$rotation)
  
  projected.metrics <- projected.metrics[,
      cluster := rownames(obj.avg)]
  
  projected.metrics <- data.table(projected.metrics)[, 
    donor := unlist(data.table(projected.metrics)[,strsplit(cluster, '_')][2,])][,
    cluster := unlist(data.table(projected.metrics)[, strsplit(cluster, '_')][1,])][]
  
  ggplot(
    projected.metrics, aes(x=PC1, y=PC2, label=cluster, color=donor)) +
    geom_point() + 
    geom_text_repel()
```

### Act_Clust Assignment:

```{r}

#
# GLYC 0 7
# OXPHOS 2 13 3
# TC2 10 14
# IFNG 4 11
# GNLY 1
# INACT 5 6 8 9 12
# REMOVE 15 16 17

#CD8 old annotated cluters
cd8_act_cluster_annotations <- fread(text="
seurat_clusters   Activation    Subtype
0         ACT         ACT-GLYC
1         ACT         ACT-GNLY
2         ACT         ACT-OXPHOS
3         ACT         ACT-OXPHOS
4         ACT         ACT-IFNG
5         INACT       INACT
6         INACT       INACT
7         ACT         ACT-GLYC
8         INACT       INACT
9         INACT       INACT
10        ACT         ACT-TC2
11        ACT         ACT-IFNG
12        INACT       INACT
13        ACT         ACT-OXPHOS
14        ACT         ACT-TC2
")

cd8_act_cluster_annotations[, seurat_clusters := factor(seurat_clusters)]
addtl_metadata <- cd8_act_cluster_annotations[data.table(scrna_cd8@meta.data), on=c('seurat_clusters')]
scrna_cd8@meta.data$Activation <- addtl_metadata$Activation
scrna_cd8@meta.data$Subtype <- addtl_metadata$Subtype
scrna_cd8@meta.data$Subtype <- factor(scrna_cd8@meta.data$Subtype)

cluster_cols = c(
  'ACT-GLYC'=brewer.pal(9,'OrRd')[8],
  'ACT-OXPHOS'=brewer.pal(9,'YlOrRd')[5],
  'ACT-GNLY'=brewer.pal(9,'YlOrRd')[3],
  'ACT-IFNG'=brewer.pal(9,'PuRd')[5],
  'ACT-TC2'=brewer.pal(9,'PuRd')[3],
  'REST'=brewer.pal(9,'YlGnBu')[6],
  'UNTR'=brewer.pal(9,'PuBu')[3],
  'REST-TCM'=brewer.pal(9,'YlGn')[6],
  'REST-CXCR3'=brewer.pal(9,'YlGn')[3])

cluster_name <- 'Subtype'
car_clust_pct <- data.table(scrna_cd8@meta.data)[
  as.numeric(get(cluster_name)) %in% c(0:12,14), .N, 
  by=c('car',cluster_name)][, 
    pct_car := N/sum(N), by='car'][,
      clust_count := sum(N), by=cluster_name][,
        log2_clust_enrich := log2(pct_car/mean(pct_car)), by=cluster_name]

car_clust_pct[, log2_ce_trunc := ifelse(
  test= abs(log2_clust_enrich) > 2, 
  yes= 2 * sign(log2_clust_enrich), 
  no= log2_clust_enrich)]

cd8_act_enrich_plot <- ggplot(car_clust_pct) + 
  geom_point(aes(
    x=get(cluster_name),
    y=factor(car, c("CD28","4-1BB","BAFF-R","TACI","CD40","CD30","TNR8","Zeta","KLRG1","Untransduced")),
    fill=log2_ce_trunc, size=pct_car), pch=21, color='black') + 
  scale_fill_distiller(palette='PRGn', direction=1, limits=c(-1,1), na.value = brewer.pal(9,'PRGn')[1]) + scale_size(range=c(1,10)) + theme_minimal()
```

### CD8 Gene Set Plots

```{r}
# oxphos vs glyc
oxphos_vs_glyc_diffs <- diff_genes_in_clusts(scrna_cd8, c(2,3,14,10), c(0,7), label_top_rna=50)
#glyc genes
message(print(paste(gsub('.adt','', oxphos_vs_glyc_diffs$diff_genes[avg_log2FC < 0 & -log(p_val_adj) > 50]$rn),collapse='\n')))
#osphos genes
message(print(paste(gsub('.adt','', oxphos_vs_glyc_diffs$diff_genes[avg_log2FC > 0 & -log(p_val_adj) > 10]$rn),collapse='\n')))

# ifng vs oxphos and glyc
ifng_diffs <- diff_genes_in_clusts(scrna_cd8, c(4,11), c(0,2,3,7,14,10), label_top_rna=50)
message(print(paste(gsub('.adt','', diffs$diff_genes[avg_log2FC > 0 & -log(p_val_adj) > 50]$rn),collapse='\n')))

# oxphos vs ifng and glyc
oxphos_diffs <- diff_genes_in_clusts(scrna_cd8, c(2,3,13), c(0,1,4,6,10,11,14), label_top_rna=50)
message(print(paste(gsub('.adt','', diffs$diff_genes[avg_log2FC > 0 & -log(p_val_adj) > 50]$rn),collapse='\n')))

# gnly vs ifng/glyc/oxphos
gnly_diffs <- diff_genes_in_clusts(scrna_cd8, c(1), c(0,2,3,4,10,11,13,14), label_top_rna=50)
message(print(paste(gsub('.adt','', diffs$diff_genes[avg_log2FC > 0 & -log(p_val_adj) > 50]$rn),collapse='\n')))



```

### CD8 Combined 

```{r}
scrna_cd8_all <- LoadH5Seurat(here::here(
  '..','..','scrnaseq_tcsl154',
  'rds',
  paste0('scrna.wnn.int.CD8.all.clean.h5Seurat')))

cd8_all_metadata <- data.table(
  scrna_cd8_all@meta.data, keep.rownames=T)[,
    list(rn_orig=rn, sample, donor, orig_t_type=t_type)][
      donor == 'TCSL154', rn := gsub('(.*)-\\d_\\d','\\1',rn_orig)]

sample_subtype_map <- merge(cd8_all_metadata, original_metadata, by='rn', all=T)[!is.na(Subtype) & !is.na(rn_orig)]
get_cluster_enrich_plot(scrna_cd8_all, 'seurat_clusters', hide='', facet_formula=formula(k_type~car))

scrna_cd8_all@meta.data[,'subtype_154'] <- factor(NA, levels=levels(sample_subtype_map$Subtype))
scrna_cd8_all@meta.data[sample_subtype_map$rn_orig,]$subtype_154 <- sample_subtype_map$Subtype

scrna_cd8_all <- map_clusters(scrna_cd8, scrna_cd8_all, 'Subtype', 'Subtype')
scrna_cd8_all <- map_clusters(scrna_cd8, scrna_cd8_all, 'seurat_clusters', 'cd8_act_clust')

int_cd8_all_subset <- subset(combined.integrated.data,
    cells=intersect(
      names(scrna_cd8_all$cell), 
      names(combined.integrated.data$cell)))

scrna_cd8_all[['RNA']] <- int_cd8_all_subset@assays$RNA
scrna_cd8_all[['ADT']] <- int_cd8_all_subset@assays$ADT


DimPlot(scrna_cd8_all, reduction = 'wnn.umap', group.by = 'subtype_154', label = TRUE, repel = TRUE, label.size = 2.5) +
  DimPlot(scrna_cd8_all, reduction = 'wnn.umap', group.by = 'seurat_clusters', label = TRUE, repel = TRUE, label.size = 2.5) +
  DimPlot(scrna_cd8_all, reduction = 'wnn.umap', group.by = 'Subtype', label = TRUE, repel = TRUE, label.size = 2.5)

# cd8_umap_plots <- umap_plot(scrna_cd8, assay='SCT_INT', cluster_name='seurat_clusters', reduction='wnn.umap')
# cd8_umap_plots_adt <- umap_plot(scrna_cd8, assay='SCT_ADT_INT', cluster_name='seurat_clusters', reduction='wnn.umap')

clust_mapping_154 <- compare_clusters(scrna_cd8_all, 'seurat_clusters', 'subtype_154')
plot_grid(clust_mapping_154$plot_a_in_b, clust_mapping_154$plot_b_in_a)

clust_mapping_154 <- compare_clusters(scrna_cd8_all, 'Subtype', 'subtype_154')
plot_grid(clust_mapping_154$plot_a_in_b, clust_mapping_154$plot_b_in_a)

clust_mapping_154 <- compare_clusters(scrna_cd8_all, 'Subtype', 'seurat_clusters')
plot_grid(clust_mapping_154$plot_a_in_b, clust_mapping_154$plot_b_in_a)


ggplot(data.table(scrna_cd8_all@meta.data)[car != 'K_only',
    list(N=.N, seurat_clusters), by=c('donor','car','k_type')][, 
      pct := .N/N, by=c('donor','car','seurat_clusters','k_type')]) + 
  geom_bar(aes(x=seurat_clusters, fill=donor, y=pct), stat='identity', position='dodge') + 
  facet_grid(car + k_type ~ .)


```

#### Clust Act Plots

```{r}

### FC Activation plots
```{r}

cd4_pct_act <- unique(map_clust_act(scrna_cd4_all, cluster_name='Subtype')$car_clust_data[
  !is.na(Subtype), list(Subtype, cd19_clust_act_l2fc)])
cd8_pct_act <- unique(map_clust_act(scrna_cd8_all, cluster_name='seurat_clusters')$car_clust_data[
  !is.na(seurat_clusters), list(seurat_clusters, cd19_clust_act_l2fc)])

act_data <- rbind(cd8_pct_act[, t_type := 'CD8'], cd4_pct_act[, t_type := 'CD4'])
act_data[cd19_clust_act_l2fc == Inf, cd19_clust_act_l2fc := max(act_data[cd19_clust_act_l2fc != Inf]$cd19_clust_act_l2fc)]

scrna_dt <- act_data[scrna_dt, on=c('t_type','Subtype')]

ggplot(scrna_dt[!is.na(Subtype)]) + 
  geom_point(aes(x=wnnUMAP_1, y=wnnUMAP_2, color=cd19_clust_act_l2fc), size=0.6) +
  coord_fixed() +
  theme_minimal() +
  scale_color_distiller('Stim. vs. Unstim', palette='PuOr', direction=-1, oob=scales::squish, limits=c(-3,3), breaks=c(-3,-1,1,3), labels=rev(c('>8x stim','2x stim','2x unstim','>8x unstim'))) +
  theme(axis.title=element_blank(),
    axis.text=element_blank(),
    axis.ticks=element_blank(), panel.grid=element_blank())
```


```


### CD4 Act Clust Assignment:

```{r}
scrna_cd4 <- LoadH5Seurat(here::here(
  '..','..','scrnaseq_tcsl154',
  'rds',
  paste0('scrna.wnn.int.CD4.act_clust.clean.h5Seurat')))

cd4_metadata <- data.table(
  scrna_cd4@meta.data, keep.rownames=T)[,
    list(rn_orig=rn, sample, donor, orig_t_type=t_type)][
      donor == 'TCSL154', rn := gsub('(.*)-\\d_\\d','\\1',rn_orig)]

scrna.wnn.cd4_all_metadata <- readRDS("~/scrnaseq_tcsl154/rds/scrna.wnn.cd4_all_metadata.rds")
scrna.wnn.cd8_all_metadata <- readRDS("~/scrnaseq_tcsl154/rds/scrna.wnn.cd8_all_metadata.rds")

original_metadata <- rbind(
    data.table(scrna.wnn.cd4_all_metadata, keep.rownames=T)[, list(rn_154=rn, sample, t_type, Subtype)],
    data.table(scrna.wnn.cd8_all_metadata, keep.rownames=T)[, list(rn_154=rn, sample, t_type, Subtype)])[,
      rn := gsub('(.*)-\\d','\\1',rn_154)]

sample_subtype_map <- merge(cd4_metadata, original_metadata, by='rn', all=T)[!is.na(Subtype) & !is.na(rn_orig)]
get_cluster_enrich_plot(scrna_cd4, 'seurat_clusters', hide='', facet_formula=formula(~car))

scrna_cd4@meta.data[,'subtype_154'] <- factor(NA, levels=levels(sample_subtype_map$Subtype))
scrna_cd4@meta.data[sample_subtype_map$rn_orig,]$subtype_154 <- sample_subtype_map$Subtype

plot_grid(
  get_cluster_enrich_plot(scrna_cd4, 'seurat_clusters', hide='', facet_formula=formula(~car))$plot,
  DimPlot(scrna_cd4, reduction = 'wnn.umap', group.by = 'subtype_154', label = TRUE, repel = TRUE, label.size = 2.5) +
    DimPlot(scrna_cd4, reduction = 'wnn.umap', group.by = 'seurat_clusters', label = TRUE, repel = TRUE, label.size = 2.5),
  ncol=1)
```