
---
title: "TCSL154 scRNA Analysis"
output:
  html_document:
    df_print: paged
knit: (function(inputFile, encoding) {
  rmarkdown::render(inputFile, encoding = encoding, output_dir = here::here('..','html'))})
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(dplyr)
library(tidyr)
library(tidyr)
library(Seurat)
library(patchwork)
library(cluster)
library(parallelDist)
library(ggplot2)
library(data.table)
library(cowplot)

reload_data=F
source(here::here('r','fig_colors.R'))
```

```{r load_data, eval=reload_data}
# Load the dataset
scrna.data <- Read10X(data.dir = here::here(
    '..','..','scrnaseq_tcsl154',
    'cellranger','outs','raw_feature_bc_matrix'))

#rename some adt features
hto_rows <- paste('Hashtag',1:12,sep='_')

adt_feature_names <- fread(here::here(
    '..','..','scrnaseq_tcsl154',
    'meta','adt_names.tsv'))
rownames(scrna.data$`Antibody Capture`) <- c(paste0(adt_feature_names$vis_name,'.adt'), hto_rows)

#rna features
scrna <- CreateSeuratObject(
    counts = scrna.data$`Gene Expression`, 
    project = "tcsl154", 
    min.cells = 3,
    min.features = 200)

adt_rows <- setdiff(rownames(scrna.data$`Antibody Capture`), hto_rows)

#adt features
scrna[['ADT']] <- CreateAssayObject(
    counts = scrna.data$`Antibody Capture`[adt_rows, colnames(scrna)])
```

## RNA QC metrics

```{r rna_qc, fig.width=4, fig.height=4, eval=reload_data}

scrna[["percent.mt"]] <- PercentageFeatureSet(scrna, pattern = "^MT-")

VlnPlot(scrna, 
  pt.size=F, 
  features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)

plot1 <- FeatureScatter(
  scrna, feature1 = "nCount_RNA", feature2 = "percent.mt")

plot2 <- FeatureScatter(
  scrna, feature1 = "nCount_RNA", feature2 = "nFeature_RNA")

plot1 + plot2

scrna <- subset(
  scrna, 
  subset = nFeature_RNA > 800 & nFeature_RNA < 7000 & percent.mt < 15)
```

## HTO QC Metrics, subsetting, clustering

```{r hto, fig.width=10, fig.height=10, eval=reload_data}

source(here::here('r','scrna','HTOdemux.R'))

scrna[['HTO']] <- CreateAssayObject(
  counts = scrna.data$`Antibody Capture`[hto_rows, colnames(scrna)])

#all data loaded, clear raw data from memory and do garbage collect
#rm('scrna.data')
gc()

scrna <- NormalizeData(scrna, assay = "HTO", normalization.method = "CLR")

scrna <- DoubleHTODemux(scrna, positive.quantile = 0.90, ncenters=25)

ggplot(
    data.table(scrna@meta.data)[, .N, 
      by=.(HTO_classification, HTO_clusters, hash.ID)],
    aes(y=HTO_classification, x=hash.ID, fill=log10(N))) + 
  geom_tile() + facet_wrap(~HTO_clusters, scales='free') +
  labs(title='Cluster vs HTO assignment')

# assign match for best cluster
cluster.hash.ID <- data.table(scrna@meta.data)[, 
  .N, by=.(HTO_classification, HTO_clusters, hash.ID)][, 
    list(cluster.hash.ID= hash.ID[hash.ID != 'Triplets'][
      which.max(N[hash.ID != 'Triplets'])]), by=.(HTO_clusters)]

updated_metadata <- cluster.hash.ID[
  data.table(scrna@meta.data), 
    on='HTO_clusters', nomatch=NA, all=T][
    is.na(cluster.hash.ID), cluster.hash.ID := 'Triplet']

updated_metadata[, HTO_best_doublet := gsub('\\+\\d+','',HTO_classification)]
updated_metadata[, HTO_miscluster := (
  as.character(cluster.hash.ID) != as.character(HTO_best_doublet))]

# margin to throw away clustering points if third-closest HTO's
# signal strength relative to second is beyond this 
HTO_margin_cutoff <- -1

ggplot(updated_metadata[
    HTO_miscluster == F & 
    (as.character(hash.ID) != as.character(cluster.hash.ID))]) + 
  geom_jitter(aes(x=HTO_margin, y=HTO_classification), alpha=0.2) + 
  facet_wrap(~cluster.hash.ID, scales='free') + 
  geom_vline(aes(xintercept=-1))

# mark cells with HTO margin above cutoff as misclustered
updated_metadata[
  (as.character(hash.ID) != as.character(cluster.hash.ID)) & 
  HTO_margin > HTO_margin_cutoff, 
    HTO_miscluster := T]

# assign cluster hashes if not misclustered
updated_metadata[HTO_miscluster == F, hash.ID := cluster.hash.ID]

# merge with metadata file, add to HTO scrna metadata
HTO_sample_metadata <- fread(here::here(
    '..','..','scrnaseq_tcsl154','meta','sample_metadata.csv'))
names(HTO_sample_metadata)[2] <- 'hash.ID'
updated_metadata <- HTO_sample_metadata[updated_metadata, on='hash.ID']

#if hash combo is not a sample, label negative
updated_metadata[
  is.na(sample_num) & !(hash.ID %in% c('Triplets', 'Singlets', 'Negative')), 
  hash.ID := 'Negative']

# update metadata in object
updated_metadata <- as.data.frame(updated_metadata)
rownames(updated_metadata) <- colnames(scrna)
scrna@meta.data <- updated_metadata

#assign new cell identities
Idents(object = scrna) <- scrna@meta.data$hash.ID

#make a subset for tsne plot
hto_subset <- subset(scrna, cells=sample(Cells(scrna), 10000))

hto.dist.mtx <- as.matrix(parDist(t(
  GetAssayData(object = hto_subset, assay = "HTO"))))

hto_subset <- RunTSNE(hto_subset, 
  distance.matrix = hto.dist.mtx,
  perplexity = 100)
```

## Analyzing RNA Data for TCSL 155

```{r initial_rna_analysis, fig.width=10, fig.height=10, eval=reload_data}
scrna.hashed <- subset(scrna, idents = HTO_sample_metadata$hash.ID)

DefaultAssay(scrna.hashed) <- "RNA"
scrna.hashed <- NormalizeData(scrna.hashed, 
    assay='RNA', 
    normalization.method = "LogNormalize", scale.factor = 10000)

scrna.hashed <- ScaleData(
  scrna.hashed, assay='RNA',
  features= VariableFeatures(scrna.hashed))
 
scrna.hashed <- FindVariableFeatures(scrna.hashed,
    selection.method = "vst",
    nfeatures = 2000)

saveRDS(scrna.hashed, file = here::here(
    '..','..','scrnaseq_tcsl154',
    'rds','scrna.hashed.rds'))
```
## Separating CD4s and CD8s

First, get adt tags that separate CD4s and CD8s, and add rnas as controls.

```{r cd4_v_cd8, fig.width=4, fig.height=4}

if (!('scrna.hashed' %in% ls()))
  scrna.hashed <- readRDS(here::here(
    '..','..','scrnaseq_tcsl154',
    'rds','scrna.hashed.rds'))

# old_idents <- Idents(scrna.hashed)
# scrna.hashed@meta.data[['car_k_type']] <- paste(
#   scrna.hashed@meta.data[['car']], 
#   scrna.hashed@meta.data[['k_type']], sep='_')
# Idents(scrna.hashed) <- 'car_k_type'
# DimPlot(scrna.hashed, label=T)

ggplot(data.table(hto_subset@meta.data)[!is.na(k_type) & !is.na(car)]) + 
  geom_bar(aes(x=car, fill=k_type)) + 
  facet_wrap(~k_type, scale='free_x') + 
  theme_minimal()

DefaultAssay(scrna.hashed) <- "ADT"
scrna.hashed <- NormalizeData(scrna.hashed, assay = "ADT", normalization.method = "CLR")
scrna.hashed <- ScaleData(scrna.hashed, assay = "ADT")
scrna.hashed <- FindVariableFeatures(scrna.hashed)

adt_features <- scrna.hashed@assays$ADT@var.features

cd4v8_subset_vars <- c(adt_features, 'rna_CD8A','rna_CD8B','rna_CD4')

cd4v8.dt <- data.table(FetchData(
  object = scrna.hashed, 
  vars = cd4v8_subset_vars))

combined_pca <- prcomp(cd4v8.dt)

# calculate pca stats
pca.dt <- data.table(pc = data.table(colnames(combined_pca$rotation))[, 
    PC := as.integer(gsub("[A-Z]", "", V1))][, PC], 
    sd = combined_pca$sdev, 
    var = combined_pca$sdev^2, 
    var.norm = combined_pca$sdev^2/sum(combined_pca$sdev^2), 
    var.acc = cumsum(combined_pca$sdev^2/sum(combined_pca$sdev^2)))

melt.pca.dt <- melt(
  pca.dt, measure.vars = c("var.norm", "var.acc"), variable.name = "metric")

pca.stats.plot <- ggplot(melt.pca.dt) +
  geom_line(aes(x = pc, y = value, color = metric)) +
  geom_point(aes(x = pc, y = value, color = metric)) +
  scale_x_continuous(limits = c(1, NA)) +
  labs(title = "Fraction of Variance Captured by Principal Components, Every Sort Group", 
       x = "Principal Component", y = "Fraction of Variance") +
  theme_bw()

pca.stats.plot
```

```{r cd4_v_cd8_pca, fig.width=6, fig.height=16, eval=reload_data}
# project data onto principal components
projected.metrics <- scale(cd4v8.dt, combined_pca$center, combined_pca$scale) %*% combined_pca$rotation

#vectorized switch:
names_4v8 <- c('4-8-', '8+4-', '4+8-', '4+8+')
categories_4v8 <- c('CD8','intermediate','intermediate','CD4')

projected.metrics <- cbind.data.frame(cd4v8.dt, projected.metrics)
projected.metrics[, is_CD4_adt := CD4.adt > 1.2]
projected.metrics[, is_CD8_adt := CD8A.adt > 0.5]
projected.metrics[, adt_state := names_4v8[1+(2*is_CD4_adt + is_CD8_adt)]]
projected.metrics[, is_CD4_rna := rna_CD4 > 0]
projected.metrics[, is_CD8B_rna := rna_CD8B > 0]
projected.metrics[, is_CD8A_rna := rna_CD8A > 0]
projected.metrics[, rna_state := names_4v8[1+(2*is_CD4_rna + (is_CD8B_rna | is_CD8A_rna))]]

intercepts <- c(.89, 1.95)
slopes <- c(.77, 1.45)

projected.metrics[, cd8_classify := PC2 > (slopes[1]*PC1 + intercepts[1])]
projected.metrics[, cd4_classify := PC2 > (slopes[2]*PC1 + intercepts[2])]
projected.metrics[, cd_category := categories_4v8[1+(2*cd4_classify + cd8_classify)]]

plot_grid(
  ggplot(projected.metrics[], aes(x = PC1, y = PC2, color=adt_state)) + 
    geom_point(alpha=0.1) + 
    geom_abline(intercept=intercepts, slope=slopes),
  ggplot(projected.metrics[], aes(x = PC1, y = PC2, color=adt_state)) + 
    facet_wrap(~adt_state) + 
    geom_point(alpha=0.1) + 
    geom_abline(intercept=intercepts, slope=slopes),
  ggplot(projected.metrics[], aes(x = PC1, y = PC2, color=adt_state)) + 
    facet_wrap(~rna_state) + 
    geom_point(alpha=0.1) + 
    geom_abline(intercept=intercepts, slope=slopes),
  ggplot(projected.metrics[], aes(x = PC1, y = PC2, color=cd_category)) + 
    geom_point(alpha=0.1) + 
    geom_abline(intercept=intercepts, slope=slopes),
  ncol=1, 
  labels=c('', 'Grouped by presence of CD4/CD8 ADT', 'Grouped by presence of CD4/CD8 RNA','Manual Classification'), 
  scale=0.9)

scrna.hashed@meta.data[['CD4v8']] <- projected.metrics$cd_category

if (reload_data) saveRDS(scrna.hashed, file = here::here(
    '..','..','scrnaseq_tcsl154',
    'rds','scrna.hashed.rds'))
```

## CiteSeq Subsetting

```{r fig.width=15, fig.height=8, unit="inches"}

if (!('scrna.hashed' %in% ls()))
  scrna.hashed <- readRDS(here::here(
    '..','..','scrnaseq_tcsl154',
    'rds','scrna.hashed.rds'))

car_order <- c('CD28','4-1BB','BAFF-R','TACI','CD40','TNR8','Zeta','KLRG1','Untransduced')
car_labels <- c('CD28','4-1BB','BAFF-R','TACI','CD40','CD30','Zeta','KLRG1','Untransduced')

car_colors <- c(
    '4-1BB'=blues[1], 'BAFF-R'=greens[1], 
    'CD28'=blues[2], 'CD40'=purples[2], 
    'KLRG1'=reds[5], 'TACI'=greens[2], 
    'CD30'=purples[1], 'Zeta'=grey)

scrna.hashed@meta.data$car <- factor(scrna.hashed@meta.data$car, levels=car_order, labels=car_labels)
scrna.cd4 <- subset(scrna.hashed, subset= (CD4v8 == 'CD4'))
scrna.cd8 <- subset(scrna.hashed, subset= (CD4v8 == 'CD8'))

make_citeseq_dt <- function(obj_subset, assay='ADT') {
  return(data.table(cbind(t(obj_subset@assays[[assay]]@data), obj_subset@meta.data))[car != 'K_only'])
}

make_rna_dt <- function(obj_subset) {
  return(data.table(cbind(t(obj_subset@assays$RNA@data), obj_subset@meta.data))[car != 'K_only'])
}

make_pct_grid <- function(
    obj_subset, x_axis, x_cutoff, y_axis, y_cutoff, split=c('car','k_type','CD4v8')) 
{
  pct_dt <- obj_subset[,
    quadrant := factor((get(x_axis) > x_cutoff)*2 + (get(y_axis) > y_cutoff), 
      levels=c(0:3), label=c('BL','TL','BR','TR'))][, 
        .N, by=c('quadrant', split)][,
          list(pct= N/sum(N), count= N, quadrant), by=split]
 
  pos_dt <- data.table(
    quadrant=c('BL','TL','BR','TR'),
    x=c(-Inf, -Inf, Inf, Inf),
    y=c(-Inf,Inf,-Inf,Inf))
 
  pct_dt <- pos_dt[pct_dt, on='quadrant']
  
  
  labels <- paste0(
    rep(gsub('(.*)\\.\\w+','\\1',x_axis),4),
    c('-','+','-','+'),
    '/',
    rep(gsub('(.*)\\.\\w+','\\1',y_axis),4),
    c('-','-','+','+'))
  
   pct_dt[, lbl := factor(quadrant, labels=labels)]

  
  return(pct_dt)
}

citeseq_density_2d <- function(obj_subset, x_axis, x_cutoff, y_axis, y_cutoff, 
    title=paste0(x_axis,' vs ',y_axis)) 
{
  ggplot(obj_subset, aes_string(x=x_axis, y=y_axis)) + 
      stat_density_2d(aes(fill = stat(ndensity)), geom = "raster", contour = FALSE) + 
      scale_fill_distiller(palette='PuRd', direction=1) + facet_grid(k_type~car) +  
      geom_hline(yintercept=y_cutoff) + geom_vline(xintercept=x_cutoff) + 
      scale_y_continuous(expand = c(0, 0)) + 
      scale_x_continuous(expand=c(0,0)) + 
      labs(title=title) +
      geom_text(
        data=make_pct_grid(obj_subset, x_axis, x_cutoff, y_axis, y_cutoff), 
        size=3, 
        aes(x=x, y=y, hjust=(x > 0), vjust=(y > 0), label=paste0(as.character(round(pct,3)*100),'%\n','(',count,')'))) +
      theme_minimal()
}

citeseq_bar_plot <- function(obj_subset, x_axis, x_cutoff, y_axis, y_cutoff, 
    title=paste0(x_axis,' vs ',y_axis))
{
  ggplot(make_pct_grid(obj_subset, x_axis, x_cutoff, y_axis, y_cutoff)) + 
      scale_fill_brewer(palette='Set1') +
      geom_bar(aes(fill=lbl, x=car, y=pct), stat='identity', position='dodge') +
      facet_wrap(~k_type, ncol=1) + theme_bw()
}

density_bar_plots <- function (obj_subset, x_axis, x_cutoff, y_axis, y_cutoff, 
    title=paste0(x_axis,' vs ',y_axis))
{
  title <- ggdraw() + draw_label(title, fontface='bold')
  plot_grid(
    title,
    citeseq_density_2d(obj_subset, x_axis, x_cutoff, y_axis, y_cutoff, title=''),
    citeseq_bar_plot(obj_subset, x_axis, x_cutoff, y_axis, y_cutoff, title=''),
    ncol=1, 
    rel_heights=c(0.2,1,1)
  )
}

#### RO v 62L
# BAFF-R has enriched 62L+/RO+ population in CD8s only
plot_grid(
  density_bar_plots(make_citeseq_dt(scrna.cd4), x_axis='CD45RO.adt', x_cutoff=0.25, y_axis='CD62L.adt', y_cutoff=0.6, title='CD4 RO v 62L'),
  density_bar_plots(make_citeseq_dt(scrna.cd8), x_axis='CD45RO.adt', x_cutoff=0.25, y_axis='CD62L.adt', y_cutoff=0.6, title='CD8 RO v 62L'),
  ncol=2
)

#### RO v CCR7
plot_grid(
  density_bar_plots(make_citeseq_dt(scrna.cd4), x_axis='CD45RO.adt', x_cutoff=0.25, y_axis='CCR7.adt', y_cutoff=0.3, title='CD4 RO v CCR7'),
  density_bar_plots(make_citeseq_dt(scrna.cd8), x_axis='CD45RO.adt', x_cutoff=0.25, y_axis='CCR7.adt', y_cutoff=0.3, title='CD8 RO v CCR7'),
  ncol=2
)

#### RO+, CD62L v CCR7
plot_grid(
  density_bar_plots(make_citeseq_dt(scrna.cd4)[CD45RO.adt > 0.25], x_axis='CD62L.adt', x_cutoff=0.6, y_axis='CCR7.adt', y_cutoff=0.25, title='CD4, RO+, 62L v CCR7'),
  density_bar_plots(make_citeseq_dt(scrna.cd8)[CD45RO.adt > 0.25], x_axis='CD62L.adt', x_cutoff=0.6, y_axis='CCR7.adt', y_cutoff=0.25, title='CD8, RO+, 62L v CCR7'),
  ncol=1
)

#### RO+, CD62L v CD27
plot_grid(
  density_bar_plots(make_citeseq_dt(scrna.cd4)[CD45RO.adt > 0.25], x_axis='CD62L.adt', x_cutoff=0.6, y_axis='CD27.adt', y_cutoff=0.4, title='CD4, RO+, 62L v CD27'),
  density_bar_plots(make_citeseq_dt(scrna.cd8)[CD45RO.adt > 0.25], x_axis='CD62L.adt', x_cutoff=0.6, y_axis='CD27.adt', y_cutoff=0.4, title='CD8, RO+, 62L v CD27'),
  ncol=2
)

#### RO+, CCR7 v CD49b
plot_grid(
  density_bar_plots(make_citeseq_dt(scrna.cd4)[CD45RO.adt > 0.25], x_axis='CCR7.adt', x_cutoff=0.25, y_axis='CD49b.adt', y_cutoff=0.6, title='CD4 CCR7 v 49b'),
  density_bar_plots(make_citeseq_dt(scrna.cd8)[CD45RO.adt > 0.25], x_axis='CCR7.adt', x_cutoff=0.25, y_axis='CD49b.adt', y_cutoff=0.6, title='CD8 CCR7 v 49b'),
  ncol=2
)
```

# PCA of CARs
```{r}

library(ggrepel)
library(cowplot)

plot_pca <- function(obj_subset, axes=c('car','k_type'), min_pct_exp = 10, assay='RNA') {
  
  obj_subset@meta.data[['pca_group']] <- do.call(paste, c(obj_subset@meta.data[axes], sep="_"))
  Idents(obj_subset) <- 'pca_group'
  
  rna_gene_set <- rownames(obj_subset@assays[[assay]]@data)[
    (tabulate(obj_subset@assays$RNA@data@i + 1) / 
      ncol(obj_subset@assays$RNA@data) * 100) > min_pct_exp]
  
  adt_gene_set <- rownames(obj_subset@assays[[assay]]@data)[
    (tabulate(obj_subset@assays$RNA@data@i + 1) / 
      ncol(obj_subset@assays$ADT@data) * 100) > min_pct_exp]
  
  obj.avg <- AverageExpression(obj_subset, assays=c('ADT','RNA'))
  obj.avg <- t(rbind(obj.avg$ADT, obj.avg$RNA))
  obj.avg <- obj.avg[, apply(obj.avg,2,sd) > 0]
  obj.avg <- obj.avg[, colnames(obj.avg) %in% c(rna_gene_set, adt_gene_set)]
  combined_pca <- prcomp(scale(obj.avg))
  
  # calculate pca stats
  pca.dt <- data.table(pc = data.table(colnames(combined_pca$rotation))[, 
      PC := as.integer(gsub("[A-Z]", "", V1))][, PC], 
      sd = combined_pca$sdev, 
      var = combined_pca$sdev^2, 
      var.norm = combined_pca$sdev^2/sum(combined_pca$sdev^2), 
      var.acc = cumsum(combined_pca$sdev^2/sum(combined_pca$sdev^2)))
  
  melt.pca.dt <- melt(
    pca.dt, measure.vars = c("var.norm", "var.acc"), variable.name = "metric")
  
  pca_metrics_plot <- ggplot(melt.pca.dt) +
    geom_line(aes(x = pc, y = value, color = metric)) +
    geom_point(aes(x = pc, y = value, color = metric)) +
    scale_x_continuous(limits = c(1, NA)) +
    labs(title = "Fraction of Variance Captured by Principal Components, Every Sort Group", 
         x = "Principal Component", y = "Fraction of Variance") +
    theme_bw()
  
  projected.metrics <- data.table(
    scale(obj.avg, combined_pca$center, combined_pca$scale) %*% combined_pca$rotation)
  
  projected.metrics <- projected.metrics[,
      car := rownames(obj.avg)]
  
  pca_plot <- ggplot(
      projected.metrics, aes(x=PC1, y=PC2, label=car)) +
    geom_point() + 
    geom_text_repel()
  
  return(
    list(pca_plot= pca_plot, combined_pca= combined_pca, pca_metrics_plot= pca_metrics_plot)
  )
}

plot_gene_assoc_bar <- function(pca_data, PC_name, n_genes=10) {
  
  gene_data <- data.table(pca_data$combined_pca$rotation, keep.rownames=TRUE)[
    c(order(get(PC_name))[1:n_genes], order(-get(PC_name))[1:n_genes]), 
    list(association= get(PC_name), gene= rn)][, 
      gene := factor(gene, levels=gene[order(association)])]

  ggplot(gene_data) + 
    geom_bar(aes(x=association, y=gene, fill=factor(sign(association))), 
      stat='identity', color='black') + 
    theme_minimal() + 
    theme(legend.position='none') +
    labs(x='',y='', title=PC_name)
}

plot_pca_cars <- function(obj_subset, title, axes=c('car','k_type'), min_pct_exp=10, assay='RNA') {
  pca_data <- plot_pca(obj_subset, axes, min_pct_exp, assay)
  return(plot_grid(
    pca_data$pca_plot + theme_minimal() + labs(title=title),
    plot_gene_assoc_bar(pca_data, 'PC1', n_genes=25),
    plot_gene_assoc_bar(pca_data, 'PC2', n_genes=25),
    ncol=3, nrow=1,
    rel_widths=c(5, 1, 1)))
}

plot_pca_cars(subset(scrna.cd4, subset=(k_type == 'cd19+' & car %in% c('4-1BB','BAFF-R','CD40','TACI','CD28','Zeta'))), title='CD4 CD19+ PCA (Stimulatory CARs)')

plot_pca_cars(subset(scrna.cd8, subset=(k_type == 'cd19+' & car %in% c('4-1BB','BAFF-R','CD40','TACI','CD28','Zeta'))), title='CD8 CD19+ PCA (Stimulatory CARs)')

plot_pca_cars(subset(scrna.hashed, subset=(k_type == 'cd19+' & car %in% c('4-1BB','BAFF-R','CD40','TACI','CD28','Zeta'))), title='CD19+ PCA (Stimulatory CARs)')

plot_pca_cars(subset(scrna.cd8, subset=(car %in% c('4-1BB','BAFF-R','CD40','TACI','CD28','Zeta', 'KLRG1','Untransduced'))), title='All CARs', axes=c('car','k_type','CD4v8'))

RidgePlot(subset(scrna.cd4, subset=(k_type == 'cd19+' & car %in% c('4-1BB','BAFF-R','CD40','TACI','CD28','Zeta'))), features = c('TMSB4X'))

cd4pos <- subset(scrna.cd4, subset=(k_type == 'cd19+' & car %in% c('4-1BB','BAFF-R','CD40','TACI','CD28','Zeta')))
Idents(cd4pos) <- 'car'
VlnPlot(cd4pos, features=c('ANKS1A','ZNF66','BCAT1'))

comp_car_volcano <- function(obj_subset, ident.1, ident.2, logfc.thresh=0.2, min.pct=0.1,
    assay='RNA', lbl_pval_cutoff = 10, title='') {
  
  title <- paste(title,
    paste('[',paste(ident.2,collapse=', '),'] <- vs -> [',paste(ident.1,collapse=', '),']'),
    sep='\n')
  
  Idents(obj_subset) <- 'car'
  markers <- FindMarkers(
    obj_subset, assay=assay,
    ident.1 = ident.1, ident.2=ident.2,
    logfc.threshold = logfc.thresh, min.pct=min.pct)
  
  volcano_ggplot <- ggplot(markers, aes(x=avg_logFC, y=-log10(p_val_adj))) + geom_point() +
    geom_hline(aes(yintercept=-log10(0.05))) + 
    scale_x_continuous(limits=c(-max(abs(markers$avg_logFC)), max(abs(markers$avg_logFC)))) +
    geom_text_repel(
      data=data.table(markers, keep.rownames = T)[-log10(p_val_adj) > lbl_pval_cutoff],
      aes(label=rn)) + 
    theme_minimal() +
    labs(x=title, y='Adjusted -log10(p)')
  
  return(list(markers=markers, plot=volcano_ggplot))
}

comp_car_volcano(subset(scrna.cd4, subset=(k_type == 'cd19+')), ident.1=c('CD28'), ident.2='BAFF-R', lbl_pval_cutoff = 3, min.pct=0.25, assay='RNA')

```

## correlation plots

```{r}
library(spatstat)
library(dendextend)
library(ggdendro)


corr_plots <- function(subset_obj, assay='ADT') {
  
  DefaultAssay(subset_obj) <- assay
  
  citeseq_cor <- make_citeseq_dt(subset_obj, assay=assay)[, 
      data.table(cor(.SD[, c(rownames(subset_obj@assays[[assay]]@data)), with=F]), keep.rownames=T)]
  
  #remove NA rows
  citeseq_cor <- na.omit(
    citeseq_cor[, which(colSums(is.na(citeseq_cor)) != (nrow(citeseq_cor) - 1)), with=F])
   
  # generate heirarchical clustering
  dendro_m <- as.matrix(citeseq_cor[, -1])
  rownames(dendro_m) <- unlist(citeseq_cor[,1])
  dendro <- as.dendrogram(hclust(d = dist(x = dendro_m)))
  dendro <- rotate(dendro, names(sort(rowMeans(dendro_m))))
  
  # row order from clustering
  citeseq_cor_melt <- melt(citeseq_cor, id.vars='rn')
  
  # melt and reorder rows
  citeseq_cor_melt$rn <- factor(
      citeseq_cor_melt$rn,
      levels = labels(dendro))
  citeseq_cor_melt$variable <- factor(
      citeseq_cor_melt$variable,
      levels = labels(dendro))
  
  # remove markers that are not strongly correlated with anything
  var_has_corr <- citeseq_cor_melt[variable != rn, max(abs(value)) > 0.075, by='variable'][V1 == T]$variable
  citeseq_cor_melt_subset <- citeseq_cor_melt[variable %in% var_has_corr & rn %in% var_has_corr]
  
  average_corr <- ggplot(copy(citeseq_cor_melt_subset)[value > 0.5, value := 0.5]) + 
    geom_tile(aes(x=rn, y=variable, fill=value)) + 
    theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) + 
    scale_fill_distiller(palette='PRGn', limits=c(-0.5, 0.5))
  
  return(average_corr)
}
```

## Gene Set Enrichment

https://learn.gencore.bio.nyu.edu/rna-seq-analysis/gene-set-enrichment-analysis/

```{r}

# BiocManager::install("clusterProfiler", version = "3.8")
# BiocManager::install("pathview")
# BiocManager::install("enrichplot")
# BiocManager::install(organism, character.only = TRUE)
library(clusterProfiler)
library(enrichplot)
library(ggplot2)
library(organism, character.only = TRUE)

gene_data <- data.table(
    comp_car_volcano(subset(scrna.cd4, subset=(k_type == 'cd19+')), 
    ident.1=c('CD28'), ident.2='BAFF-R', lbl_pval_cutoff = 3, min.pct=0.25, 
    assay='RNA')$markers, 
    keep.rownames = T)
names(gene_data)[1] <- 'gene'

gene_list <- gene_data$avg_logFC
names(gene_list) <- gene_data$gene
gene_list = sort(gene_list, decreasing = TRUE)

organism = "org.Hs.eg.db"
gse <- gseGO(geneList=gene_list, 
             ont ="MF", 
             keyType = "SYMBOL", 
             nPerm = 10000, 
             minGSSize = 3, 
             maxGSSize = 100, 
             pvalueCutoff = 0.05, 
             verbose = TRUE, 
             OrgDb = organism, 
             pAdjustMethod = "none")

dotplot(gse, showCategory=25, split='.sign') + facet_grid(.~.sign)
cnetplot(gse, categorySize="pvalue", foldChange=gene_list, showCategory = 50)

ids<-bitr(names(gene_list), fromType = "SYMBOL", toType = "ENTREZID", OrgDb=organism)
#some manual ones:
ids <- data.table(
    rbind(
    ids,
    c('SEPT6', 23157),
    c('WARS', 7453),
    c('H2AFZ', 3015),
    c('RARRES3', 5920),
    c('H3F3A', 3020),
    c('FAM96B', 51647),
    c('GARS', 2617),
    c('H1FX', 8971),
    c('KIAA1551',55196),
    c('EPRS', 2058),
    c('AES',166)))
names(ids) <- c('gene','entrez')

gene_data_kegg <- data.table(ids)[gene_data, on='gene']

# Create a vector of the gene unuiverse
kegg_gene_list <- gene_data_kegg$avg_logFC
# Name vector with ENTREZ ids
names(kegg_gene_list) <- gene_data_kegg$entrez

# omit any NA values 
kegg_gene_list<-na.omit(kegg_gene_list)

# sort the list in decreasing order (required for clusterProfiler)
kegg_gene_list = sort(kegg_gene_list, decreasing = TRUE)

kegg_organism = "hsa"
kk2 <- gseKEGG(geneList     = kegg_gene_list,
               organism     = kegg_organism,
               nPerm        = 10000,
               minGSSize    = 3,
               maxGSSize    = 800,
               pvalueCutoff = 0.05,
               pAdjustMethod = "none",
               keyType       = "ncbi-geneid")

kk2@result$core_enrichment <- gene_data_kegg[
    data.table(kk2@result)[, list(
        entrez=unlist(strsplit(core_enrichment, '/'))), by='ID'], 
            on='entrez'][, list(core_enrichment=paste(gene, collapse='/')), by='ID']$core_enrichment



dotplot(kk2, showCategory = 10, title = "Enriched Pathways" , split=".sign") + facet_grid(.~.sign)

kegg_gene_list <- gene_data_kegg$avg_logFC
# Name vector with ENTREZ ids
names(kegg_gene_list) <- gene_data_kegg$gene
# omit any NA values 
kegg_gene_list<-na.omit(kegg_gene_list)
# sort the list in decreasing order (required for clusterProfiler)
kegg_gene_list = sort(kegg_gene_list, decreasing = TRUE)

cnetplot(kk2, categorySize="pvalue", foldChange=kegg_gene_list, showCategory=50)

```
## Make UMAP plots and objects for each subset

```{r umap_each_subset, eval=reload_data}

umap_plot <- function(subset, plot_title_text="", assay, cluster_resolution) {

  subset.markers <- FindAllMarkers(subset,
  features = VariableFeatures(subset),
  min.pct = 0.5, logfc.threshold = 0.25)
  
  top.subset.markers <- subset.markers %>% group_by(cluster) %>% top_n(n = 10, wt = avg_logFC)
  
  top_genes_plot <- ggplot(top.subset.markers) + 
    geom_bar(aes(y=avg_logFC, x=reorder(gene, avg_logFC)), stat='identity') + 
    facet_wrap(~cluster, scales='free', ncol=3) +
    theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))
  
  cluster_col_name <- paste0(assay,'_snn_res.',as.character(cluster_resolution))
  
  cluster_pct_plot <- ggplot(
      data.table(subset@meta.data)[
        CD4v8 != 'intermediate', .N, 
        by=c('car', cluster_col_name, 'CD4v8', 'k_type')][,
          list(cluster=get(cluster_col_name), t_type=CD4v8, frac=N/sum(N)), 
          by=c('car', 'CD4v8', 'k_type')]) + 
    geom_bar(aes(x=car, y=frac, fill=frac), stat='identity', color='grey50') +
    scale_fill_distiller(palette='YlGnBu', direction=1) +
    facet_grid(k_type+t_type~cluster) + 
    theme_bw() +
    coord_flip()
  
  plot_title <- ggdraw() + 
    draw_label(
      plot_title_text,
      fontface = 'bold',
      x = 0,
      hjust = 0
    ) +
    theme(
      # add margin on the left of the drawing canvas,
      # so title is aligned with left edge of first plot
      plot.margin = margin(0, 0, 0, 7)
    )
  
  gene_heatmap <- DoHeatmap(subset,
        features = unique(
          (top.subset.markers %>% group_by(cluster) 
            %>% top_n(n=5, wt= avg_logFC))$gene))
      
  umap_plots <- plot_grid(
    plot_grid(
      plot_title,
      DimPlot(subset, reduction = "umap", label=T),
      cluster_pct_plot,
      rel_heights=c(0.2,1.25,1),
      ncol=1),
    gene_heatmap,
    ncol=2)
  
  return(umap_plots)
}

umap_plot_by_car <- function(subset, plot_title_text="", assay, cluster_resolution) {

  Idents(subset) <- 'car'
  subset.markers <- FindAllMarkers(subset,
  features = VariableFeatures(subset),
  min.pct = 0.5, logfc.threshold = 0.25)
  
  top.subset.markers <- subset.markers %>% group_by(cluster) %>% top_n(n = 10, wt = avg_logFC)
  
  top_genes_plot <- ggplot(top.subset.markers) + 
    geom_bar(aes(y=avg_logFC, x=reorder(gene, avg_logFC)), stat='identity') + 
    facet_wrap(~cluster, scales='free', ncol=3) +
    theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))
  
  cluster_col_name <- paste0(assay,'_snn_res.',as.character(cluster_resolution))
  
  cluster_pct_plot <- ggplot(
      data.table(subset@meta.data)[
        CD4v8 != 'intermediate', .N, 
        by=c('car', cluster_col_name, 'CD4v8', 'k_type')][,
          list(cluster=get(cluster_col_name), t_type=CD4v8, frac=N/sum(N)), 
          by=c('car', 'CD4v8', 'k_type')]) + 
    geom_bar(aes_string(x='cluster', y='frac', fill='frac'), stat='identity', color='grey50') +
    scale_fill_distiller(palette='YlGnBu', direction=1) +
    facet_grid(k_type+t_type~car) + 
    theme_bw() +
    coord_flip()
  
  plot_title <- ggdraw() + 
    draw_label(
      plot_title_text,
      fontface = 'bold',
      x = 0,
      hjust = 0
    ) +
    theme(
      # add margin on the left of the drawing canvas,
      # so title is aligned with left edge of first plot
      plot.margin = margin(0, 0, 0, 7)
    )
  
  gene_heatmap <- DoHeatmap(subset,
        features = unique(
          (top.subset.markers %>% group_by(cluster) 
            %>% top_n(n=5, wt= avg_logFC))$gene))
  
  car_density_map <- ggplot(
      cbind(subset@meta.data, Embeddings(object = subset[["umap"]]))) + 
    geom_density2d(aes(x=UMAP_1, y=UMAP_2)) + facet_wrap(~car) +
    theme_minimal()
      
  umap_plots <- plot_grid(
    plot_grid(
      plot_title,
      car_density_map,
      cluster_pct_plot,
      rel_heights=c(0.2,1.25,1),
      ncol=1),
        plot_grid(
      gene_heatmap,
      top_genes_plot,
      ncol=1,
      rel_heights=c(3,2)),
    ncol=2)
  
  return(umap_plots)
}

umap_subset <- function(this_subset, 
    plot_title_text="", assay='RNA', cluster_resolution=0.8,
    save=T, save_plots=T, rerun=F, return_all=F) {
  
  DefaultAssay(this_subset) <- assay
  this_subset_umap <- subset(this_subset, 
    features=rownames(this_subset@assays[[assay]]))
  
  subset_name <- gsub(' ','.',plot_title_text)
  
  rds_file <- here::here(
    '..','..','scrnaseq_tcsl154',
    'rds',
    paste0(
        'scrna.hashed.',
        subset_name,
        '.rds'))
  
  plot_dir <- here::here(
    '..','figs','scrna',
    'subset_plots',
    subset_name)
  
  if (save_plots) dir.create(plot_dir, recursive=T)
  
  if (rerun || !file.exists(rds_file)) {
    
      if (assay == 'RNA') {
        this_subset_umap <- NormalizeData(
            this_subset_umap,
            normalization.method = "LogNormalize", 
            scale.factor = 10000)
      } else if (assay == 'ADT') {
          this_subset_umap <- NormalizeData(
            this_subset_umap,
            normalization.method = "CLR")
      }
      
      this_subset_umap <- FindVariableFeatures(
          this_subset_umap, 
          selection.method = "vst",
          nfeatures = 2000)
      
      this_subset_umap <- ScaleData(
        this_subset_umap, assay=assay,
        features= VariableFeatures(this_subset_umap))
    
      this_subset_umap <- RunPCA(
        this_subset_umap, features = VariableFeatures(this_subset_umap))
    
      this_subset_umap <- FindNeighbors(this_subset_umap, reduction = "pca", dims = 1:8)
      this_subset_umap <- FindClusters(this_subset_umap, resolution = cluster_resolution, algorithm=1)
      this_subset_umap <- RunUMAP(this_subset_umap, dims = 1:8)
    
      this_subset_umap@meta.data[['car.ttype.ktype']] <- paste(
        this_subset_umap@meta.data[['car']],
        this_subset_umap@meta.data[['CD4v8']],
        this_subset_umap@meta.data[['k_type']], sep='_')
  } else {
      this_subset_umap <- readRDS(rds_file)
  }
  
  # plots
  subset_umap_plot <- umap_plot(this_subset_umap, plot_title_text, assay, cluster_resolution)
  car_umap_plot <- umap_plot_by_car(this_subset_umap, plot_title_text, assay, cluster_resolution)
  car_pca_plot <- plot_pca_cars(this_subset, title=plot_title_text, axes=c('car','k_type','CD4v8'))
  
  if (save_plots) {
    fn_pref = paste0(plot_dir,'/',subset_name,'.')
    ggsave(paste0(fn_pref,'subset_umap.pdf'), subset_umap_plot, width=15, height=9)
    ggsave(paste0(fn_pref,'car_umap_plot.pdf'), car_umap_plot, width=15, height=9)
    ggsave(paste0(fn_pref,'car_pca_plot.pdf'), car_pca_plot, width=10, height=10)
  }
  
  # volcano plots
  if (assay == 'RNA') {
    Idents(this_subset) <- 'car'
    a_sets = list(c('BAFF-R'), c('TACI'), c('CD40'), c('BAFF-R','TACI'), c('BAFF-R','TACI','CD40'))
    b_sets = list(c('CD28'), c('CD28','Zeta'), c('CD28','4-1BB','Zeta'))
    volcano_plots = list()
    
    for (ident_a in a_sets) {
      volcano_plots[[ident_a]] = list()
      ident_a_name <- paste(ident_a, collapse='_')
      for (ident_b in b_sets) {
        ident_b_name <- paste(ident_b, collapse='_' )
        volcano_plots[[ident_a_name]][[ident_b_name]] <- comp_car_volcano(
          this_subset, ident.1=ident_a, ident.2=ident_b, lbl_pval_cutoff = 3, min.pct=0.25, assay=assay)
        if (save_plots) {
          fn = paste0(plot_dir,'/',subset_name,'.',
            ident_a_name,'_VS_',ident_b_name,'.volcano.pdf')
          ggsave(fn, volcano_plots[[ident_a_name]][[ident_b_name]]$plot, width=8, height=8, units='in')
        }
      }
    }
  }
  
  if ((rerun | (!file.exists(rds_file))) & save)
    saveRDS(this_subset, file = rds_file)
    
  if (return_all) {
    return(list(
      cluster_plot=subset_umap_plot,
      car_plot=car_umap_plot,
      volcano_plots=volcano_plots,
      pca_plot=car_pca_plot,
      rds_file=rds_file))
  } else {
    return(list(
        cluster_plot=subset_umap_plot,
        car_plot=car_umap_plot,
        volcano_plots=volcano_plots,
        pca_plot=car_pca_plot,
        rds_file=rds_file))
  }
}

stimulatory_cars <- c('4-1BB','BAFF-R','CD40','TACI','CD28','Zeta')

if (!reload_data && !('umap_plots' %in% ls()) )
{
  umap_plots <- readRDS(here::here(
    '..','..','scrnaseq_tcsl154',
    'rds','scrna.umap.plots.rds')) 
} else {
  subset_fxns= list(
      
      # all cars
      function(obj) subset(obj, subset = (car != 'K_only')),
      function(obj) subset(obj, subset = (CD4v8 == 'CD4' & car != 'K_only')),
      function(obj) subset(obj, subset = (CD4v8 == 'CD8' & car != 'K_only')),
      function(obj) subset(obj, subset = (k_type == 'cd19+' & car != 'K_only')),
      
      # stimulated 4s and 8s
      function(obj) subset(obj, subset = (k_type == 'cd19+' & CD4v8 == 'CD4' & car != 'K_only')),
      function(obj) subset(obj, subset = (k_type == 'cd19+' & CD4v8 == 'CD8' & car != 'K_only')),
      
      
      #unstimulated 4s and 8s
      function(obj) subset(obj, subset = (k_type == 'baseline' & car != 'K_only')),
      function(obj) subset(obj, subset = (k_type == 'baseline' & CD4v8 == 'CD4' & car != 'K_only')),
      function(obj) subset(obj, subset = (k_type == 'baseline' & CD4v8 == 'CD8' & car != 'K_only')),
      
      #stimulatory cars only
      function(obj) subset(obj, subset = (k_type == 'cd19+' & CD4v8 == 'CD4' & car %in% stimulatory_cars)),
      function(obj) subset(obj, subset = (k_type == 'cd19+' & CD4v8 == 'CD8' & car %in% stimulatory_cars)),
  
      #stimulatory cars, central memory
      function(obj) subset(obj, subset = (k_type == 'cd19+' & CD4v8 == 'CD4' & car %in% stimulatory_cars &
        CD45RO.adt > 0.25 & CD62L.adt > 0.6)),
      function(obj) subset(obj, subset = (k_type == 'cd19+' & CD4v8 == 'CD8' & car %in% stimulatory_cars &
        CD45RO.adt > 0.25 & CD62L.adt > 0.6)),
      #stimulatory cars, effector memory
      function(obj) subset(obj, subset = (k_type == 'cd19+' & CD4v8 == 'CD4' & car %in% stimulatory_cars &
        CD45RO.adt > 0.25 & CD62L.adt < 0.6)),
      function(obj) subset(obj, subset = (k_type == 'cd19+' & CD4v8 == 'CD8' & car %in% stimulatory_cars &
        CD45RO.adt > 0.25 & CD62L.adt < 0.6))
    )


  subset_names= c('All','CD4','CD8','CD19',
                  'CD4.CD19','CD8.CD19',
                  'baseline','CD4.baseline', 'CD8.baseline',
                  'CD4.CD19.stimulatory','CD8.CD19.stimulatory',
                  'CD4.CD19.stimulatory.tcm','CD8.CD19.stimulatory.tcm',
                  'CD4.CD19.stimulatory.teff','CD8.CD19.stimulatory.teff')
  
  umap_plots <- data.table(subset_name= subset_names, subset_fxn_num=1:length(subset_fxns))
  
  # umap_plots[, ADT := list(list(
  #     umap_subset(subset_fxns[[subset_fxn_num]](scrna.hashed), plot_title_text=paste0(.BY[1],' ADT'), assay='ADT'))),
  #     by='subset_name']
  umap_plots[subset_fxn_num >=10, RNA := list(list(
      umap_subset(subset_fxns[[subset_fxn_num]](scrna.hashed), plot_title_text=paste0(.BY[1],' RNA'), assay='RNA'))), 
      by='subset_name']

  
  saveRDS(umap_plots, file = here::here(
      '..','..','scrnaseq_tcsl154',
      'rds','scrna.umap.plots.rds'))
}

```

## Show UMAP Plots

### ADT plots

```{r, fig.width=18, fig.height=10}

if (!('umap_plots' %in% ls()))
  umap_plots <- readRDS(here::here(
    '..','..','scrnaseq_tcsl154',
    'rds','scrna.umap.plots.rds'))

umap_plots[, print(ADT[[1]]), by='subset_name']
```

### RNA plots

```{r, fig.width=18, fig.height=10}

umap_plots[, print(RNA[[1]]$cluster_plot), by='subset_name']
```